test/e2e/volume_provisioning.go

```diff
@@ -0,0 +1,197 @@
+	// Requested size of the volume
+	requestedSize = "1500Mi"
+	// Expected size of the volume is 2GiB, because all three supported cloud
+	// providers  allocate volumes in 1GiB chunks.
+	expectedSize = "2Gi"
+)
+
+var _ = Describe("Dynamic provisioning of persistent volumes ", func() {
+	framework := NewFramework("volume-provisioning")
+
+	// filled in BeforeEach
+	var c *client.Client
+	var ns string
+
+	BeforeEach(func() {
+		c = framework.Client
+		ns = framework.Namespace.Name
+	})
+
+	////////////////////////////////////////////////////////////////////////
+	// Cinder
+	////////////////////////////////////////////////////////////////////////
+
+	Describe("DynamicProvisioner", func() {
+		It("should create and delete persistent volumes", func() {
+			SkipUnlessProviderIs("openstack", "gce", "aws")
+			By("creating a claim with a dynamic provisioning annotation")
+			claim := createClaim(ns)
+			defer func() {
+				c.PersistentVolumeClaims(ns).Delete(claim.Name)
+			}()
+			claim, err := c.PersistentVolumeClaims(ns).Create(claim)
+			Expect(err).NotTo(HaveOccurred())
+
+			err = waitForPersistentVolumeClaimPhase(api.ClaimBound, c, ns, claim.Name, 1*time.Second, 30*time.Second)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("checking the claim")
+			// Get new copy of the claim
+			claim, err = c.PersistentVolumeClaims(ns).Get(claim.Name)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Get the bound PV
+			pv, err := c.PersistentVolumes().Get(claim.Spec.VolumeName)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Check sizes
+			expectedCapacity := resource.MustParse(expectedSize)
+			pvCapacity := pv.Spec.Capacity[api.ResourceName(api.ResourceStorage)]
+			Expect(pvCapacity.Value()).To(Equal(expectedCapacity.Value()))
+
+			requestedCapacity := resource.MustParse(requestedSize)
+			claimCapacity := claim.Spec.Resources.Requests[api.ResourceName(api.ResourceStorage)]
+			Expect(claimCapacity.Value()).To(Equal(requestedCapacity.Value()))
+
+			// Check PV properties
+			Expect(pv.Spec.PersistentVolumeReclaimPolicy).To(Equal(api.PersistentVolumeReclaimDelete))
+			expectedAccessModes := []api.PersistentVolumeAccessMode{api.ReadWriteOnce}
+			Expect(pv.Spec.AccessModes).To(Equal(expectedAccessModes))
+			Expect(pv.Spec.ClaimRef.Name).To(Equal(claim.ObjectMeta.Name))
+			Expect(pv.Spec.ClaimRef.Namespace).To(Equal(claim.ObjectMeta.Namespace))
+
+			// We start two pods:
+			// - The first writes 'hello word' to the /mnt/test (= the volume).
+			// - The second one runs grep 'hello world' on /mnt/test.
+			// If both suceed, Kubernetes actually allocated something that is
+			// persistent across pods.
+			By("checking the created volume is writable")
+			runInPodWithVolume(c, ns, claim.Name, "echo 'hello world' > /mnt/test/data")
+
+			By("checking the created volume is readable and retains data")
+			runInPodWithVolume(c, ns, claim.Name, "grep 'hello world' /mnt/test/data")
+
+			// Ugly hack: if we delete the AWS/GCE/OpenStack volume here, it will
```

Comment:
Not for this PR. But this reminds me that we need to think through the logic that triggers deletion of auto-provisioned volumes. Opened up https://github.com/kubernetes/kubernetes/issues/22975